File: 2015/05/src/main.rs
use fancy_regex::Regex as Rgx;
use regex::Regex;
use std::fs;

fn read_file(filepath: &str) -> Vec<String> {
    let contents = fs::read_to_string(filepath);
    let binding = contents.expect("REASON");
    let lines = binding
        .split('\n')
        .filter(|c| c.len() > 0)
        .map(|c| c.to_string())
        .collect();
    lines
}

fn nice_string(lines: &Vec<String>) -> u64 {
    let mut total: u64 = 0;
    let vowels = Regex::new(r"a|e|i|o|u").unwrap();
    let pairs = Rgx::new(r"([a-z])\1").unwrap();
    let forb = Regex::new(r"ab|cd|pq|xy").unwrap();
    for line in lines {
        let nr_vowels = vowels.captures_iter(&line).count();
        let exists_pairs = pairs.is_match(&line).unwrap();
        let nr_forb = forb.captures_iter(&line).count();
        if nr_vowels >= 3 && exists_pairs && nr_forb <= 0 {
            total += 1;
        }
    }
    total
}

fn nicer_string(lines: &Vec<String>) -> u64 {
    let mut total: u64 = 0;
    let pairs = Rgx::new(r"([a-z]{2}).*\1").unwrap();
    let middle = Rgx::new(r"([a-z])[a-z]\1").unwrap();
    for line in lines {
        let is_pair = pairs.is_match(&line).unwrap();
        let is_middle = middle.is_match(&line).unwrap();
        if is_pair && is_middle {
            total += 1;
        }
    }
    total
}

fn part1() {
    let lines = read_file("INPUT");
    println!("Part 1: {}", nice_string(&lines));
}

fn part2() {
    let lines = read_file("INPUT");
    println!("Part 2: {}", nicer_string(&lines));
}

fn main() {
    println!("Year 2015 day 5 - Doesn't He Have Intern-Elves For This?");
    part1();
    part2();
}

File: 2015/05/main.exs
IO.puts("Year 2015 day 5 - Doesn't He Have Intern-Elves For This?")

if {:ok, string} = File.read("INPUT") do
	result = String.trim(string)
	|> String.split("\n")
	|> Enum.filter(fn x -> String.length(x) > 0 end)
	|> Enum.filter(fn x -> String.match?(x, ~r/([a-z])\1/) end)
	|> Enum.filter(fn x -> String.graphemes(x) |> Enum.filter(fn x -> x == "a" or x == "e" or x == "i" or x == "o" or x == "u" end) |> Enum.count() >= 3 end)
	|> Enum.filter(fn x -> not String.match?(x, ~r/(?:ab)|(?:cd)|(?:pq)|(?:xy)/) end)
	|> Enum.count()
	IO.puts("Part 1: #{result}")
end

if {:ok, string} = File.read("INPUT") do
	result = String.trim(string)
	|> String.split("\n")
	|> Enum.filter(fn x -> String.length(x) > 0 end)
	|> Enum.filter(fn x -> String.match?(x, ~r/([a-z]{2}).*\1/) end)
	|> Enum.filter(fn x -> String.match?(x, ~r/([a-z]).\1/) end)
	|> Enum.count()
	IO.puts("Part 2: #{result}")
end
